### ISSUE TITLE
about KOOPMAN matrix

### ISSUE THREAD

**Issue by RuiGangzai** [NickGeneva/blog-code#2]

**Comment by NickGeneva** [comment]
Hi @RuiGangzai 

Sorry for the late response. When I was working on this I found that this form seemed to work better for the system in the blog but this doesn't necessarily extrapolate to other systems. I would suggest starting with a fully parameterized matrix and then slowly adding constraints to it empirically. This could be a block diagonal form, symmetric properties, etc. Typically I have found a fully Koopman matrix tends to overfit, finding high-frequency modes.

In my more recent paper, on page 10, there's some references to different deep Koopman work with different biases on the Koopman matrix.

[https://arxiv.org/pdf/2010.03957.pdf](https://arxiv.org/pdf/2010.03957.pdf)

**Comment by RuiGangzai** [comment]
Hi Nick, thank you very much for your answers! Recently I tried to apply it to other systems, and your code helped me a lot. But there are still two problems that haunt me：For a new system, how to design the weight of each item in the loss function? In addition, how to design the number of hidden layers and the number of neurons？Looking forward to your reply, it is really important for me, thank you !

**Comment by NickGeneva** [comment]
Hi @RuiGangzai ,

Those are both great questions that I don't have complete answers for. However here was my experience from playing around with deep Koopman.

 > For a new system, how to design the weight of each item in the loss function?
 
 Indeed this is tricky, and usually, I found weighting the auto-encoder loss higher than the Koopman loss component was important. But this is entirely empirical. What really matters is controlling the number of time-steps you predict in a single loss (so how many Koopman matrix multiplications you apply). I found that you always want to start with just 1-2 steps and maybe slowly increase the number of steps as the model learns.
 
 > In addition, how to design the number of hidden layers and the number of neurons？
 
Not too much thought went into this for this work. Basically just running the model and testing a few different configurations to see what works best. Ideally, you can just specify a large enough network that is somewhat over-parameterized and the optimization will produce something that works. For example, I may just randomly start with a 4-layer model with 256 neuron layers because these are just nice numbers that work on the hardware I am training on. At the end of the day, this is a problem that all DL practitioners need to make a judgment call for any problem.

**Comment by RuiGangzai** [comment]
Hi Nick, thank you very much for your answer. Recently I am systematic study of deep learning, after I will re-study your paper, look forward to communicate and discuss with you, thank you！

**Comment by RuiGangzai** [comment]
Hi Nick, sorry to bother you again.
In your paper you mentioned above, I noticed that you used 2D-CNN and 3D-CNN in the embedding model of 2D fluid dynamics and 3D Reaction–diffusion dynamics.  
I mean, you used a convolutional network instead of a fully connected network such as the embedding model structure of the Lorenz system. So I have the following question:
In the embedding model of Lorenz system or the blog code of duffing equation, can 1D-CNN be used instead of fully connected network? If so, how should we do it？
I hope you can convert the blog code into 1D-CNN form. I want to study it carefully because I am still a beginner. If you do I really very grateful! Thanks a million.

**Comment by RuiGangzai** [comment]
Hi@NickGeneva, for data with large differences, we usually use standardization for processing and retraining, but this will cause the trained Koopman matrix to no longer reflect the dynamic behavior of the original system. Have you encountered similar problems ？Thank you !

**Comment by RuiGangzai** [comment]
Hi @NickGeneva, for data with large differences, we usually use standardization for processing and retraining, but this will cause the trained Koopman matrix to no longer reflect the dynamic behavior of the original system. If you do not standardize the data, it will cause great difficulties in training. Have you encountered similar problems  What are the good suggestions  Thank you !

**Comment by RuiGangzai** [comment]
Hi @NickGeneva ,
How to calculate the eigenfunctions of Koopman operator according to the eigenvector of Koopman operator?
From viz(.py), my understanding from the code is ：g(x)*eigenvector of Koopman operator -> eigenfunctions of Koopman operator  -> the amplitude and phase angle
Looking forward to your answer, thank you！！
